docs/user_manual.md

```diff
@@ -49,40 +60,341 @@ python -m virtualenv venv
+├── .datumaro/
+│   ├── cache/ # object cache
+│   │   ├── <2 leading symbols of obj hash>/
+│   │   │   └── <rest symbols of obj hash>/
+│   │   │       └── <object data>
+│   ├── models/
+│   ├── plugins/ # custom project-specific plugins
+│   │   ├── plugin1/
+│   │   |   ├── __init__.py
+│   │   |   └── file2.py
+│   │   ├── plugin2.py
+│   │   └── ...
+│   ├── tmp/ # temp files
+│   └── tree/ # working tree metadata
+│       ├── config.yml
+│       └── sources/
+│           ├── <source name 1>.dvc
+│           ├── <source name 2>.dvc
+│           └── ...
+│
+├── <source name 1>/
+│   └── <source data>
+└── <source name 2>/
+    └── <source data>
+```
+
+### Use cases
+
+Consider few diagrams describing what Datumaro does for you behind the scene.
+
+Imagine we have a project with 2 data sources. Roughly, it corresponds to
+the following set of commands:
+
+```bash
+datum create
+datum add <...> -n source1
+datum add <...> -n source2
+datum transform <...> source1 # used 3 times
+datum transform <...> source2 # used 5 times
 ```
 
-As a python library:
+Then, for some reasons, the project cache was cleaned from `source1` revisions.
+We also don't have anything in the project working directories.
+Let's see what happens, if we call the `diff` command with 2 different
+revisions now.
 
-``` python
-import datumaro
+![use case 1](images/usecase1_diag.svg)
+
+Datumaro needs to reproduce 2 dataset revisions requested so that they could
+be read and compared. `source1` will be looked for in the project cache.
+It will be found missing, since the cache was cleaned (1). Then, Datumaro
+will look for previous source revisions and won't find any (2, 3). Then, it
+will try to download the original data and reproduce the resulting dataset,
+re-applying all the operations from the source history (4, 5). When it is
+done, the resulting dataset is returned (6).
+
+The `source2` will be looked for the same way. In our case, it will be found
+in the cache and returned. Once both datasets are restored and read, they
+are compared.
+
+Consider the second situation. Let's try to `export` the `source1`. Suppose
+we have a clear project cache and the `source1` has a copy in the working
+directory.
+
+![use case 2](images/usecase2_diag.svg)
+
+Again, Datumaro needs to reproduce a dataset revision (stage) requested. It
+looks for the dataset in the working directory and finds some data (1). The
+data hash is computed and compared with the saved one in the history (2).
+If the hashes match, the dataset is read and returned (3a). Otherwise,
+Datumaro tries to reproduce the dataset using the approach described above.
+Then, the dataset is exported in the requested format.
+
+To sum up, Datumaro tries to restore a dataset from the project cache or
+reproduce it from sources. It can be done as long as the source operations
+are recorded and any step data is available. Note that cache objects share
+common files, so if there are only annotation changes between datasets,
```

Comment:
```suggestion
common files, so if there are only annotation differences between datasets,
```

"Changes" implies that one dataset was obtained by modifying the other, which won't be true in the general case.